presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaMetadata.java

```diff
@@ -269,22 +272,38 @@ private boolean isTablesEnumeratingTable(SchemaTableName schemaTableName)
         Session session = ((FullConnectorSession) connectorSession).getSession();
 
         Optional<Set<String>> tables = filterString(constraint, TABLE_NAME_COLUMN_HANDLE);
+        Stream<QualifiedTablePrefix> prefixStream;
         if (tables.isPresent()) {
-            return prefixes.stream()
+            prefixStream = prefixes.stream()
                     .flatMap(prefix -> tables.get().stream()
                             .map(table -> new QualifiedObjectName(catalogName, prefix.getSchemaName().get(), table)))
                     .filter(objectName -> metadata.getTableHandle(session, objectName).isPresent() || metadata.getView(session, objectName).isPresent())
-                    .map(QualifiedObjectName::asQualifiedTablePrefix)
-                    .collect(toImmutableSet());
+                    .map(QualifiedObjectName::asQualifiedTablePrefix);
         }
+        else {
+            prefixStream = prefixes.stream()
+                    .flatMap(prefix -> Stream.concat(
+                            metadata.listTables(session, prefix).stream(),
+                            metadata.listViews(session, prefix).stream()))
+                    .filter(objectName -> !predicate.isPresent() || predicate.get().test(asFixedValues(objectName)))
+                    .map(QualifiedObjectName::asQualifiedTablePrefix);
+        }
+        return collectMaxPrefixes(prefixStream);
+    }
 
-        return prefixes.stream()
-                .flatMap(prefix -> Stream.concat(
-                        metadata.listTables(session, prefix).stream(),
-                        metadata.listViews(session, prefix).stream()))
-                .filter(objectName -> !predicate.isPresent() || predicate.get().test(asFixedValues(objectName)))
-                .map(QualifiedObjectName::asQualifiedTablePrefix)
-                .collect(toImmutableSet());
+    private Optional<ImmutableSet<QualifiedTablePrefix>> collectMaxPrefixes(Stream<QualifiedTablePrefix> prefixStream)
```

Comment:
> I like the streams api, so I just moved the limit and collection into a normal for loop, and put that in a separate method.

I also like streams, but here IMO it went too far. It looks overly complicated while it is actually not that hard. Can you please give it a try to rewrite to `else` block of `calculatePrefixesWithTableName` to regular `for-each` loop?